verifier/sources/verifier_channel.move

```diff
@@ -0,0 +1,244 @@
+module verifier_addr::verifier_channel {
+    use std::vector::{append, borrow, borrow_mut, enumerate_ref, slice};
+    use aptos_std::aptos_hash::keccak256;
+
+    use lib_addr::bytes::{bytes32_to_u256, num_to_bytes_le, vec_to_bytes_le};
+    use lib_addr::prime_field_element_0::{fmul, from_montgomery};
+    use lib_addr::vector::set_el;
+    use verifier_addr::prng::{get_random_bytes, init_prng};
+
+    // This line is used for generating constants DO NOT REMOVE!
+    // 3
+    const EMASK_MUST_BE_LESS_THAN_2_TO_THE_POWER_OF_64: u64 = 0x3;
+    // 100
+    const EOVERFLOW_PROTECTION_FAILED: u64 = 0x64;
+    // 2
+    const EPROOF_OF_WORK_CHECK_FAILED: u64 = 0x2;
+    // 113078212145816603762751633895895194930089271709401121343797004406777446400
+    const K_MONTGOMERY_R_INV: u256 = 0x40000000000001100000000000012100000000000000000000000000000000;
+    // End of generating constants!
+
+    friend verifier_addr::stark_verifier_7;
+    friend verifier_addr::fri_statement_verifier_7;
+
+    public(friend) inline fun get_prng_ptr(channel_ptr: u64): u64 {
+        channel_ptr + 1
+    }
+
+    public(friend) fun init_channel(ctx: &mut vector<u256>, channel_ptr: u64, public_input_hash: u256) {
+        // set `ctx[channel_ptr]` as index 0 in `proof`
+        set_el(ctx, channel_ptr, 0);
+        init_prng(ctx, channel_ptr + 1, public_input_hash);
+    }
+
+    // Sends a field element through the verifier channel.
+    //
+    // Note that the logic of this function is inlined in many places throughout the code to reduce
+    // gas costs.
+    public(friend) fun send_field_elements(
+        ctx: &mut vector<u256>,
+        channel_ptr: u64,
+        n_elements: u64,
+        target_ptr: u64
+    ) {
+        assert!(n_elements < 0x1000000, EOVERFLOW_PROTECTION_FAILED);
+        let bound = 0xf80000000000020f00000000000000000000000000000000000000000000001fu256;
+        let digest = *borrow(ctx, channel_ptr + 1);
+
+        for (i in target_ptr..(target_ptr + n_elements)) {
```

Comment:
let counter = *borrow(ctx, channel_ptr + 2);
        let end_ptr = target_ptr + n_elements;
        for (i in target_ptr..end_ptr) {
            let field_element = bound;
            while (field_element >= bound) {
                // keccak256(abi.encodePacked(digest, counter));
                field_element = bytes32_to_u256(keccak256(vec_to_bytes_le(&vector[digest, counter])));
                // *counterPtr += 1;
                counter = counter + 1;
            };
            // *targetPtr = fromMontgomery(fieldElement);
            set_el(ctx, i, fmul(field_element, K_MONTGOMERY_R_INV));
        };
        *borrow_mut(ctx, channel_ptr + 2) = counter;